### PR TITLE
Add AudioWebSocket close event

### PR DESCRIPTION
**Pull request by freyacodes**
Solves #11 

Adds a `Flux` stream to the API for handling close events. The flux uses a generic interface, so that other types of events can be added in the future.

This is a non-breaking change.

### CODE REVIEW

File: src/main/java/space/npstr/magma/events/api/WebSocketClosedEvent.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    public Member getMember() {
+        return member;
+    }
+
+    public int getCloseCode() {
+        return closeCode;
+    }
+
+    @Nullable
+    public CloseCode getCloseCodeEnum() {
```
Comment (schnapster) [inline src/main/java/space/npstr/magma/events/api/WebSocketClosedEvent.java]: `CloseCode` is part of the internal Magma Api. What is the case for exposing it here and therefore making it public?

File: src/main/java/space/npstr/magma/Magma.java [hunk 2]
```diff
@@ -58,6 +61,9 @@
     private static final int DEFAULT_POOL_BUFFER_SIZE = 16 * 1024;
 
     private final FluxSink<LifecycleEvent> lifecycleSink;
+    @Nullable
```
Comment (schnapster) [inline src/main/java/space/npstr/magma/Magma.java:64]: Can't we avoid this (misleading) annotation by moving the initialization of those two fields into the constructor?

File: src/main/java/space/npstr/magma/Magma.java [hunk 3]
```diff
@@ -92,6 +104,12 @@
     @Override
     public void shutdown() {
         this.lifecycleSink.next(Shutdown.INSTANCE);
+        if (apiEventSink != null) apiEventSink.complete();
```
Comment (schnapster) [inline src/main/java/space/npstr/magma/Magma.java]: Nit: Please reference class member fields with `this.`

File: src/main/java/space/npstr/magma/Magma.java [hunk 4]
```diff
@@ -92,6 +104,12 @@
...
+    }
+
+    @Override
+    public Flux<MagmaEvent> getEventStream() {
+        return apiEventFlux;
```
Comment (schnapster) [inline src/main/java/space/npstr/magma/Magma.java]: Nit: Please reference class member fields with `this.`

File: src/main/java/space/npstr/magma/events/api/WebSocketClosedEvent.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import space.npstr.magma.events.audio.ws.CloseCode;
+
+import javax.annotation.Nullable;
+
+/**
+ * This event is fired when an audio web socket is closed. However, this event is not fired if we are trying to
+ * resume (i.e reconnect) automatically unless resuming causes a new socket to close.
+ */
+@SuppressWarnings("unused")
+public class WebSocketClosedEvent implements MagmaEvent {
```
Comment (schnapster) [inline src/main/java/space/npstr/magma/events/api/WebSocketClosedEvent.java]: Please implement this class via immutables.org, like all the other value objects/events in this project.

**Review by schnapster — CHANGES_REQUESTED** [review]
Thank you for contributing this feature.
I like the general architectural choices that you have taken here, this PR looks very acceptable if you can iron out a few things. Please see my individual comments for more.

File: src/main/java/space/npstr/magma/events/api/WebSocketClosedEvent.java (hunk 1 above)
Comment (freyacodes) [reply src/main/java/space/npstr/magma/events/api/WebSocketClosedEvent.java]: In hindsight it's not very useful

File: src/main/java/space/npstr/magma/Magma.java (hunk 2 above)
Comment (freyacodes) [reply src/main/java/space/npstr/magma/Magma.java:64]: Then we would have to subscribe to the Flux. I'm not sure if subscribing prematurely is bad practice in our particular case.

**Comment by freyacodes** [comment]
Completed all the requested changes besides the one about initializing `lifecycleSink`

**Pull request merged**